Feat: Crew 테스트를 제외한 CRUD API 구현

ffinn92 opened the pull request:
### ✔️이슈번호
  -  Issue #10

### 📚작업목록
- [x] Crew 생성 API 구현
- [x] Crew 상세조회 API 구현 
- [x] Crew 조회 API 구현
- [x] Crew 수정 API 구현
- [x] Crew 삭제 API 구현
- [x] 각 API별 Java 기본 exception 구현
- [x] 각 API별 Valid 처리 구현  

### 📖작업내용
  - Crew 도메인의 기본 CRUD를 구현하였으나, Custom Exception작업과 Test 코드 작성 작업이 필요합니다.

### 🙋‍♂️참고사항
  - Embedded 값들은 공통된 값 2개 이상이 아닌경우는 모두 수정하였습니다.

ffinn92 committed:
Refactor: Embeddable 애노테이션, 값타입 제거

- 둘 이상의 공통적인 성격이 존재할때만 Embeddable 사용하기위함

ffinn92 committed:
Feat: Crew 전체 조회 API 구현

- Dto List를 data로 묶기 위한 CommonDtoList 구현

ffinn92 committed:
Feat: 특정 id로 크루 상세 조회 API 구현

- TODO 항목 추가 및 메소드 컨벤션 수정

ffinn92 committed:
Feat: 크루 상세조회 API Dto 추가 및 수정

ffinn92 committed:
Refactor: searchCrewById 메소드의 가독성을 위해 파라미터 수정

ffinn92 committed:
Refactor: CrewSearchByIdResponse 이름 수정

ffinn92 committed:
Feat: name으로 Crew 조회하는 API 구현

ffinn92 committed:
Refactor: Member Entity의 값타입 제거 및 변수 타입 수정

ffinn92 committed:
Feat: Member Entity의 Crew 관련 값 Embedded처리

- Member의 Crew관련 정보를 수정할 editCrewInfo 메서드 구현

ffinn92 committed:
Feat: Crew 생성 API 구현

- Member의 CrewInfo 정보 변경을 적용시키기 위해 CrewService 내 MemberRepository 추가

ffinn92 committed:
Fix: Crew 생성시 BaseTime에 null값 입력 오류해결

ffinn92 committed:
Feat: Crew 정보 수정 API 구현

ffinn92 committed:
Feat: Crew 생성 시 크루장 인원 추가하기 위해 생성자 수정

- CrewInfo에 TODO 항목 기입

ffinn92 committed:
Feat: MemberCount type 변경 Long -> Integer

ffinn92 committed:
Feat: Crew 삭제 API 구현

ffinn92 committed:
Feat: Crew 삭제 API 구현

- CrewRepositoy, MemberRepository 필요한 쿼리 메소드 추가 구현
- CrewService의 valid 메소드 추가 구현

ffinn92 committed:
Feat: Entity 생성자 추가 및 Member 클래스 내 inbodies 맵핑 추가

ffinn92 committed:
Feat: Crew Valid 애노테이션 추가

- Request DTO값의 검증을 위해 추가

NB993 commented on src/main/java/com/calefit/crew/CrewRepository.java:
```diff
@@ -0,0 +1,17 @@
+public interface CrewRepository extends JpaRepository<Crew, Long> {
+
+    @Query("SELECT c FROM Crew c WHERE c.name =?1")
+    Crew findCrewByName(String crewName);
```
Optional로 감싸서 반환하는 건 어떨까요?

ffinn92 replied:
Null 체크하면서 Exception 발생시킬 수 있으니 적합하겠네요.

NB993 commented on src/main/java/com/calefit/crew/CrewRepository.java:
```diff
@@ -0,0 +1,17 @@
+    Crew findCrewByName(String crewName);
+
+    @Query("SELECT name FROM Crew WHERE name LIKE '%:crewName%'")
+    List<Crew> findCrewsByName(@Param("name") String crewName);
```
data jpa에서 필드명으로 메서드 만들 수 있게 도와주는 기능을 쓰면 어노테이션 없이 `List<Crew> findByName(String crewName)`로 바로 만들 수 있지 않을까요?

ffinn92 replied:
넵 지금 그렇게 해보니 쿼리를 명시해줬을때랑 JPA에서 자동으로 만들어주는 메서드로 호출할때랑 쿼리가 동일하네요. ㅎㅎ 수정하겠습니다.

NB993 commented on src/main/java/com/calefit/crew/CrewService.java:
```diff
@@ -0,0 +1,126 @@
+    @Transactional(readOnly = true)
+    public CommonDtoList searchCrews() {
+        List<Crew> searchCrews = crewRepository.findAll();
+        validCrewsIsEmpty(searchCrews);
```
`validXXX`대신 `validateXXX`는 어떠신가요? valid라고 적혀있어서 저는 `유효한 크루들~`이라고 읽히네요 😅

ffinn92 replied:
validate가 적합해보이네요. 수정하겠습니다. :)

NB993 commented on src/main/java/com/calefit/crew/CrewService.java:
```diff
@@ -0,0 +1,126 @@
+    @Transactional(readOnly = true)
+    public CrewDetailedResponse searchCrewById(Long crewId) {
+
+        return CrewDetailedResponse.from(validIsCrewExist(crewId));
```
저도 entity -> dto로 변환시 `.from()` 형태로 통일하겠습니다 :)

ffinn92 replied:
넵 다만, 상황에 따라 정적 팩토리 메소드 네이밍 컨벤션에 맞춰서 쓰는게 좋을거 같습니다.
https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/

NB993 commented on src/main/java/com/calefit/crew/CrewService.java:
```diff
@@ -0,0 +1,126 @@
+    @Transactional(readOnly = true)
+    public CommonDtoList searchCrewByName(String crewName) {
+        List<Crew> searchCrews = crewRepository.findCrewsByName(crewName);
+        validCrewsIsEmpty(searchCrews);
```
크루 검색 결과가 없을 때 예외를 던지지 않고 size가 0인 CommonDtoList를 리턴하는 방식은 어떻게 생각하시나요? 검색 결과가 존재하지 않는 게 예외 상황은 아니라는 생각이 들었습니다.

ffinn92 replied:
요청은 `성공` 했으나, 값이 없을 뿐이니 예외처리를 하기보단 0으로 리턴하자는 말씀이시군요.
그러면 이런경우에 HTTP 상태코드 또한 200으로 응답하기보단 요청은 성공했지만 리턴하는 값이 없다 라는 의미에 204코드로 응답하는게 적합할거같은데 어떻게 생각하시나요?

이렇게 되면 Controller단에서 CommonDtoList의 size가 0일때 204를 응답하게끔 수정해야될거같습니다.

NB993 replied:
204 코드를 언제 사용하는지 검색해보니 PUT 등 요청 처리 후에 클라이언트가 데이터를 리턴받을 필요가 없는 경우, 문서를 이동할 필요가 없는 경우에 사용한다고 하네요. 저희가 검색 결과가 0인 경우도 요청 처리 성공으로 취급한다고 하면 200이 더 적합하지 않을까 생각합니다..!

* https://www.rfc-editor.org/rfc/rfc7231#section-6.3.5
* https://developer.mozilla.org/ko/docs/Web/HTTP/Status/204
* https://laeshiny.medium.com/get-%EC%9A%94%EC%B2%AD-%EC%8B%9C-%EB%8D%B0%EC%9D%B4%ED%84%B0%EA%B0%80-%EC%97%86%EC%9C%BC%EB%A9%B4-200-or-404-4ab7430084af

ffinn92 replied:
넵 알겠습니다. 응답코드에 대해서는 말이 너무 달라서 저희가 어떻게 약속하느냐에 따라 정하면될거 같습니다. 그럼 저희는 검색 결과가 0이긴 하지만 0 자체도 응답해주는 값으로 보고, 요청 성공으로 취급하여 200으로 할까요?

NB993 replied:
넵 저는 그렇게 하면 좋을 것 같습니다 :D

NB993 commented on src/main/java/com/calefit/crew/CrewService.java:
```diff
@@ -0,0 +1,126 @@
+
+    @Transactional(readOnly = true)
+    public CommonDtoList searchCrews() {
+        List<Crew> searchCrews = crewRepository.findAll();
```
`searchCrews` -> `searchedCrews`는 어떠신가요? 저는 변수명에는 `~ed`를 붙이는 게 조금 더 자연스럽게 읽히는 것 같습니다 🙂

ffinn92 replied:
넵 앞으로도 변수명에는 `~ed` 붙이도록 하겠습니다.

NB993 commented on src/main/java/com/calefit/crew/dto/CrewCreateRequest.java:
```diff
@@ -0,0 +1,25 @@
+@Getter
+public class CrewCreateRequest {
+
+    @NotBlank(message = "요청자의 식별자는 필수 입니다.")
```
Long타입에는 `@NotBlank`를 사용할 수 없을 거예요. 그래서 저는 `@NotNull`을 사용했습니닷.

ffinn92 replied:
그렇군요! 알려주셔서 감사합니다.

NB993 commented on src/main/java/com/calefit/crew/entity/Crew.java:
```diff
@@ -1,41 +1,79 @@
+    }
+
+    public Crew(Long id,
+                Boolean deleted,
```
래퍼 클래스 대신 `boolean`으로 주는 건 어떠신가요? 처음 크루가 생성됐을 때 deleted 플래그에 `Null`대신 `false`가 들어가는 게 좋지 않을까 생각합니다.

ffinn92 replied:
세세하게 놓친부분까지 체크해주셔서 감사합니다.

NB993 commented on src/main/java/com/calefit/crew/entity/Crew.java:
```diff
@@ -1,41 +1,79 @@
+                String description,
+                String image,
+                Integer memberCount,
+                Long score,
```
score도 0점부터 시작한다고 하면 래퍼클래스 대신 기본자료형으로 선언하면 어떨까요?

NB993 commented on src/main/java/com/calefit/crew/entity/Crew.java:
```diff
@@ -1,41 +1,79 @@
+        this.image = image;
+        this.memberCount = memberCount;
+        this.score = score;
+        this.members = members;
```
크루에 매핑된 members, crewApplications는 영한님 표현에 따르면 연관관계 주인의 반대편에 있는 가짜 매핑이니까 생성자로 넣지 않아도 될 것 같습니다..! 변수에 new ArrayList<>();로 이미 생성이 되어있기도 하구요.

NB993 commented on src/main/java/com/calefit/member/entity/Member.java:
```diff
@@ -50,4 +41,39 @@ public class Member extends BaseTime {
+                  Crew crew,
+                  List<Schedule> schedules,
+                  List<Template> templates,
+                  List<Inbody> inbodies) {
```
Member도 Crew와 마찬가지로 `@OneToMany` 변수들을 생성자로 넣어주지 않아도 될 것 같습니다..!

NB993 commented on src/main/java/com/calefit/member/entity/Member.java:
```diff
@@ -50,4 +41,39 @@ public class Member extends BaseTime {
+        this.inbodies = inbodies;
+    }
+
+    public void addCrew(Crew crew) {
```
크루를 생성하는 크루장 뿐만 아니라 크루에 가입하는 유저도 같이 사용하게 될 것 같아서 Role 인자도 받아야 하지 않을까요?

ffinn92 replied:
맞아요 크루원이 가입할때도 쓰이려면 해당인자를 하는게 맞지만 추후 개발할때 수정하려고했었거든요 ㅎㅎ
피오가 체크해주신김에 수정해놓도록 하겠습니다. :)

ffinn92 committed:
Refactor: Crew, Member 생성자 수정, 필드 타입 수정

- 가짜 맵핑 엔티티 생성자 제거
- 래퍼 클래스였던 deleted, score 타입 기본자료형으로 수정
- validation build.gradle에 추가

ffinn92 committed:
Feat: Long 타입에 맞는 Valid 애노테이션 수정

ffinn92 committed:
Feat: Crew 조회시 결과값이 없을 경우 예외 미처리

- 결과값 0으로 반환하기 위해 조회결과 0에 대한 예외처리를 하지 않는다.
- findCrewByName 메소드의 유연성을 위해 Optional 추가
- valid 메소드명 -> validate 메소드로 수정

NB993 commented:
고생 많으셨습니다!  👍
